AK/PrintfImplementation.h

```diff
@@ -283,166 +285,227 @@ ALWAYS_INLINE int print_signed_number(PutChFunc putch, char*& bufptr, int number
     return print_number(putch, bufptr, number, left_pad, zero_pad, field_width);
 }
 
-template<typename PutChFunc>
-ALWAYS_INLINE int printf_internal(PutChFunc putch, char* buffer, const char*& fmt, va_list ap)
+struct State {
+    bool left_pad = false;
+    bool zero_pad = false;
+    bool dot = false;
+    unsigned field_width = 0;
+    bool has_fraction_length = false;
+    unsigned fraction_length = 6;
+    unsigned long_qualifiers = 0;
+    bool size_qualifier = false;
+    bool alternate_form = 0;
+    bool always_sign = false;
+};
+
+template<typename PutChFunc, typename ArgumentListT, template<typename T, typename U = ArgumentListT> typename NextArgument>
+struct PrintfImpl {
```

Comment:
Perhaps all these arguments you pass to every function (put char callback, buffer pointer, number of written bytes, modifiers state, argument list) could be made into fields in this class?

Reply:
I'd like not to store the putch function anywhere, storing the argument list is a no-no.
My preference with this series of `Impl` structs is to keep them merely a collection of functions (they're not static because even naming the parent class in a child is a hard thing to do, since the type of `NextArgument` is...not that easy to name.

Reply:
My main concern is that all these methods accept lots of arguments, and all those arguments are the same, *and* some of them clearly come from state maintained somewhere else (such as buffer pointer and nwritten). Would be more clear to make fields of this type a canonical location for all that state.

You're not really storing the argument list, you're storing the getter function.

Reply:
> `buffer pointer`

That's _your own_ state, i.e. the caller's.

> `nwritten`

This can go into the state, that makes sense to me.

> You're not really storing the argument list, you're storing the getter function

You mean `NextArgument`? that needs to be instantiated where needed, I can't store any `NextArgument` anywhere

Reply:
Actually, a `ModifierState` is only valid for the duration of _one_ format call, so `nwritten` cannot go there either.
hmmm

Reply:
> That's your own state, i.e. the caller's.

I'm basically suggesting to unify the caller (`printf_internal`) and this callee type

Reply:
but the entire point is to keep `printf_internal` completely separate from this type...?

Reply:
Yeah...

It's just that whenever I see too many arguments being shoveled in and out of functions, my brain goes "perhaps these could be members in a class, and the functions would be the methods".

Can we get the best of both worlds somehow?

Reply:
I guess the buffer pointer and the nwritten args can be members of the class, not sure if it's possible to keep a reference to `PutChFunc` (for _any_ PutChFunc), but other two - sure.
`ap` and `state` can't be members by necessity though

Reply:
Now the `format_*` functions take two args, I doubt it can get any better without storing `ap` somewhere, which is...not something I'm comfortable with.

Reply:
I'm gonna go ahead and assume that this is ok now.